// simple-parallax-js/dist/react/simpleParallax.es.js
import V, { useState as o, useRef as H, useCallback as y, useEffect as f } from "react";
class d {
  constructor() {
    this.callbacks = /* @__PURE__ */ new Set(), this.animationFrameId = null;
  }
  static getInstance() {
    return d.instance || (d.instance = new d()), d.instance;
  }
  register(t) {
    this.callbacks.add(t), this.animationFrameId === null && this.start();
  }
  unregister(t) {
    this.callbacks.delete(t), this.callbacks.size === 0 && this.animationFrameId !== null && this.stop();
  }
  start() {
    const t = () => {
      this.callbacks.forEach((n) => n()), this.animationFrameId = requestAnimationFrame(t);
    };
    this.animationFrameId = requestAnimationFrame(t);
  }
  stop() {
    this.animationFrameId !== null && (cancelAnimationFrame(this.animationFrameId), this.animationFrameId = null);
  }
}
const S = d.getInstance(), A = (e) => {
  const [t, n] = o(0), [a, i] = o(!1), r = H(null), s = y(() => {
    r.current && n(r.current.height);
  }, []);
  return f(() => {
    const u = () => {
      i(!0), s();
    }, c = r.current;
    return c && (c.complete ? u() : c.addEventListener("load", u)), window.addEventListener("resize", s), () => {
      c && c.removeEventListener("load", u), window.removeEventListener("resize", s);
    };
  }, [e, s]), [r, t, a];
}, j = (e, t) => e * t - e, q = (e, t) => Math.ceil(e / 100 * t - t / 2), B = (e, t) => {
  const { top: n, height: a } = e, i = -a, r = t;
  if (n < i)
    return 0;
  if (n > r)
    return 100;
  const s = (n - i) / (r - i) * 100;
  return Math.min(Math.max(s, 0), 100);
}, G = (e, t) => {
  switch (t) {
    case "up":
      return `0, ${e}px, 0`;
    case "right":
      return `${-e}px, 0, 0`;
    case "down":
      return `0, ${-e}px, 0`;
    case "left":
      return `${e}px, 0, 0`;
    case "up left":
      return `${e}px, ${e}px, 0`;
    case "up right":
      return `${-e}px, ${e}px, 0`;
    case "down left":
      return `${e}px, ${-e}px, 0`;
    case "left right":
      return `${-e}px, ${-e}px, 0`;
    default:
      return `0, ${e}px, 0`;
  }
}, O = () => {
  const [e, t] = o(null);
  return f(() => {
    const n = () => {
      t(window.innerHeight);
    };
    if (typeof window < "u")
      return t(window.innerHeight), window.addEventListener("resize", n), () => {
        window.removeEventListener("resize", n);
      };
  }, []), e;
}, W = ({
  isLoaded: e,
  imageHeight: t,
  scale: n,
  boundingClientRect: a,
  orientation: i,
  maxTransition: r
}) => {
  const [s, u] = o(0), [c, w] = o(0), [g, x] = o(""), [p, b] = o(0), h = O();
  return f(() => {
    if (!e)
      return;
    const l = j(t, n);
    u(l);
  }, [e, t, n]), f(() => {
    if (!e || !h || !a)
      return;
    let l = B(a, h);
    r && (l = Math.min(l, 100 - r)), b(l);
  }, [e, a, h]), f(() => {
    const l = q(p, s);
    w(l);
  }, [p, s]), f(() => {
    const l = G(c, i);
    x(l);
  }, [c, i]), g;
}, Y = (e = {}) => {
  const [t, n] = o(!1), a = H(null);
  return f(() => {
    const i = new IntersectionObserver(
      (s) => {
        const [u] = s;
        u && (u.isIntersecting ? n(!0) : n(!1));
      },
      {
        ...e
      }
    ), { current: r } = a;
    return r && i.observe(r), () => {
      r && i.unobserve(r);
    };
  }, [e]), [a, t];
};
var D = "/Users/geoffrey/Desktop/perso/simpleParallax.js/src/react/index.tsx";
const _ = ({
  delay: e = 0.4,
  orientation: t = "up",
  scale: n = 1.4,
  overflow: a = !1,
  transition: i = "cubic-bezier(0,0,0,1)",
  maxTransition: r = null,
  children: s
}) => {
  var F;
  const u = t ?? "up", c = n ?? 1.2, w = (F = s == null ? void 0 : s.props) == null ? void 0 : F.src, [g, x] = o(!1), [p, b] = o(0), [h, l] = o(null), [C, k] = o(""), [T, z] = o(""), [I, L, P] = A(w), [M, E] = Y({
    root: null,
    rootMargin: "0px",
    threshold: Array.from(Array(101).keys(), (m) => m / 100)
  }), R = W({
    isLoaded: P,
    imageHeight: L,
    scale: c,
    boundingClientRect: h,
    orientation: u,
    maxTransition: r
  }), $ = y(() => {
    var m;
    if (!(!E && g) && (window.scrollY !== p || !g)) {
      const v = (m = I.current) == null ? void 0 : m.getBoundingClientRect();
      v && l(v), g || x(!0), b(window.scrollY);
    }
  }, [p, E, I]);
  f(() => {
    let m = `translate3d(${R})`;
    a || (m += ` scale(${c})`), k(m);
  }, [R, c]), f(() => {
    !i || !e || z(`transform ${e}s ${i}`);
  }, [i, e]), f(() => (S.register($), () => {
    S.unregister($);
  }), [$]);
  const N = V.isValidElement(s) ? V.cloneElement(s, {
    style: {
      ...s.props.style ?? {},
      transform: C,
      willChange: "transform",
      transition: T
    },
    ref: I
  }) : null;
  return /* @__PURE__ */ V.createElement("div", { ref: M, style: {
    overflow: a ? "visible" : "hidden"
  }, __self: void 0, __source: {
    fileName: D,
    lineNumber: 92,
    columnNumber: 5
  } }, N);
};
export {
  _ as default
};
